pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,179 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeclaim
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"regexp"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/samber/lo"
+	"knative.dev/pkg/logging"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter/pkg/apis/v1beta1"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const tagPeriod = 1 * time.Second
+
+type Controller struct {
+	kubeClient    client.Client
+	ec2           *ec2.EC2
+	defaultTagRgx *regexp.Regexp
+	timer         *time.Timer
+}
+
+func NewController(kubeClient client.Client, sess *session.Session) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		ec2:           ec2.New(sess),
+		defaultTagRgx: regexp.MustCompile(corev1beta1.NodePoolLabelKey + "/.*"),
+		timer:         time.NewTimer(tagPeriod),
+	}
+}
+
+func (c *Controller) Name() string {
+	return "ec2-instance-tagger"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	nodeClaim := &corev1beta1.NodeClaim{}
```

Comment:
We should filter out cases here where:
1. The NodeClaim has an annotation that it has already been tagged
2. The NodeClaim doesn't have its `status.nodeName` value resolved yet (so the value is equal to `""`